### PR TITLE
Add property tests for invariants

### LINKED ISSUE #16: Property tests for invariants
I would like to add some tests that check that structure invariants of the tree are not violated, if that sounds ok.
Functions `<|`, `><`, `insertAt`, `deleteAt` seem to be violating invariants.

### PR DESCRIPTION
**Pull request by meooow25**
These invariants are based on my understanding of RRB-trees. There seem to be a few variations of RRB-trees from what I could find online, so correct me if any of these do not apply to this implementation. Also, let me know if I missed some invariant that could be tested.

The added tests detect that:
* `<|` can create an `Unbalanced` with `m+1` children
* `><` can create a root `Unbalanced` with 1 child
* `insertAt` can create a root `Unbalanced` with 1 child
* `deleteAt` can create a root `Unbalanced` with 1 child

Side note: I found that shrinking sometimes seems to be taking forever but I haven't attempted to address it in this PR.

Fixes #16

### CODE REVIEW

**Comment by meooow25** [comment]
@konsumlamm what do you think? Are these bugs?

**Comment by konsumlamm** [comment]
Yes, I'm pretty sure, though I didn't have enough time to investigate them yet. The failures for `insertAt` and `deleteAt` are probably not separate bugs, since they use `><`.

**Comment by konsumlamm** [comment]
Those issues should now be fixed. Could you rebase, so that the CI runs again?

**Comment by meooow25** [comment]
Nice, thanks for fixing!

File: src/Data/RRBVector/Internal/Debug.hs [hunk 1]
```diff
@@ -85,3 +88,113 @@ pattern Leaf :: Array a -> Tree a
...
+-- | Structural invariants a vector is expected to hold.
+data Invariant
+    = RootSizeGt0      -- Root: Size > 0
+    | RootShiftDiv     -- Root: The shift at the root is divisible by blockShift
+    | RootSizeCorrect  -- Root: The size at the root is correct
+    | RootGt1Child     -- Root: The root has more than 1 child if not a Leaf
+    | BalShiftGt0      -- Balanced: Shift > 0
+    | BalNumChildren   -- Balanced: The number of children is blockSize unless
+                       -- the parent is unbalanced or the node is on the right
+                       -- edge in which case it is in [1,blockSize]
```
Comment (konsumlamm) [inline src/Data/RRBVector/Internal/Debug.hs:101]: I think the number of children should still be `blockSize` if the parent is unbalanced (except when the node is on the right edge ofc).

File: src/Data/RRBVector/Internal/Debug.hs [hunk 2]
```diff
@@ -85,3 +88,113 @@ pattern Leaf :: Array a -> Tree a
...
+
+validTree :: NodeDesc -> Shift -> Tree a -> Either Invariant ()
+validTree desc sh (RRB.Balanced arr) = do
+    assert BalShiftGt0 $ sh > 0
+    assert BalNumChildren $ case desc of
+        Bal -> n == blockSize
+        BalRightEdge -> n >= 1 && n <= blockSize
+        Unbal -> n >= 1 && n <= blockSize
+    assert BalFullChildren $
+        all (\t -> countElems t == 1 `unsafeShiftL` sh) expectedFullChildren
```
Comment (konsumlamm) [inline src/Data/RRBVector/Internal/Debug.hs]:
```suggestion
        all (\t -> countElems t == 1 `shiftL` sh) expectedFullChildren
```
We don't need maximum performance here, so let's be safe that we don't invoke UB.

File: src/Data/RRBVector/Internal/Debug.hs [hunk 3]
```diff
@@ -85,3 +88,113 @@ pattern Leaf :: Array a -> Tree a
...
+        Bal -> n == blockSize
+        BalRightEdge -> n >= 1 && n <= blockSize
+        Unbal -> n >= 1 && n <= blockSize
+  where
+    n = length arr
+
+-- | Check whether an Unbalanced node could be Balanced.
+couldBeBalanced :: Shift -> A.Array (Tree a) -> PrimArray Int -> Bool
+couldBeBalanced sh arr sizes =
+   all (\i -> getSize sizes i == 1 `unsafeShiftL` sh) [0 .. n-2] &&
```
Comment (konsumlamm) [inline src/Data/RRBVector/Internal/Debug.hs]:
```suggestion
   all (\i -> getSize sizes i == 1 `shiftL` sh) [0 .. n-2] &&
```

File: src/Data/RRBVector/Internal/Debug.hs (hunk 1 above)
Comment (meooow25) [reply src/Data/RRBVector/Internal/Debug.hs:101]: That doesn't seem to be the case. For example,
```hs
λ> showTree (0 <| fromList [1..256])
"Root {size = 257, shift = 8, tree = Unbalanced [Balanced [Leaf [0]],Balanced [Leaf [1,2,3,...,15,16], <...14 more leaves...> ,Leaf [241,242,...,255,256]]] ([1,257])}"
```
The newly consed leaf's parent is balanced but has only one child.

Comment (konsumlamm) [reply src/Data/RRBVector/Internal/Debug.hs:101]: You're right. I was thinking about something like
```hs
Root {size = ..., shift = 8, tree = Unbalanced [Balanced [Leaf [0],Leaf[1]], Balanced [...] ([2,...])}
```
but that's already covered by `BalNumChildren`.

**Comment by konsumlamm** [comment]
Great work, thank you very much!

**Pull request merged**